component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,322 @@
+    }
 
-        String tenantDomain = oAuthAppDO.getUser().getTenantDomain();
-        //validate signature
-        if (!isValidSignature(signedJWT, tenantDomain, jwtSubject)) {
-            return logAndReturnFalse("Signature or Message Authentication invalid for:" + jwtSubject);
+    //"REQUIRED. iss. This MUST contain the client_id of the OAuth Client."
+    private boolean validateIssuer(String issuer, OAuthAppDO oAuthAppDO) throws OAuthClientAuthnException {
+        //check whether the issuer is client_id
+        if (isEmpty(validIssuer)) {
+            return validateWithClientId(issuer, oAuthAppDO);
+        } else if (isNotEmpty(validIssuer) && !validIssuer.equals(issuer)) {
+            String errorMessage = "Invalid field :" + issuer + " is found in the JWT. It should be equal to the: " +
+                    validIssuer;
+            if (log.isDebugEnabled()) {
+                log.debug(errorMessage);
+            }
+            throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
         }
-        String validAud = getValidAudience(tenantDomain);
-
-        //validate audience, exp time, not before time, age of token, custom claims
-        return validateAudience(validAud, audience) &&
-                validateExpirationTime(expirationTime, currentTimeInMillis, timeStampSkewMillis) &&
-                checkNotBeforeTime(notBeforeTime, currentTimeInMillis, timeStampSkewMillis) &&
-                validateAgeOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis) &&
-                validateCustomClaims(claimsSet.getCustomClaims());
+        return true;
     }
 
-    /**
-     * Get the OauthAPP
-     *
-     * @param jwtSubject
-     * @return
-     * @throws IdentityOAuth2Exception
-     */
-    private OAuthAppDO getOAuthAppDO(String jwtSubject) throws IdentityOAuth2Exception {
-        OAuthAppDO oAuthAppDO = null;
-        try {
-            oAuthAppDO = OAuth2Util.getAppInformationByClientId(jwtSubject);
-        } catch (InvalidOAuthClientException e) {
-            handleException("Error while retrieving OAuth application with provided JWT information with subject:" +
-                    jwtSubject, e);
+    private Boolean validateWithClientId(String jwtClaim, OAuthAppDO oAuthAppDO) throws OAuthClientAuthnException {
+
+        if (oAuthAppDO == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Unable to find OAuth application for provided JWT claim: " + jwtClaim);
+            }
+            throw new OAuthClientAuthnException("The issuer or the subject of the assertion is invalid.", OAuth2ErrorCodes.INVALID_REQUEST);
         }
-        return oAuthAppDO;
+        String consumerKey = oAuthAppDO.getOauthConsumerKey();
+        if (!jwtClaim.equals(consumerKey)) {
+            String errorMessage = "Invalid field: " + jwtClaim + " is found in the JWT. It should be equal to the: " +
+                    consumerKey;
```

Comment:
should we send client_id in the error message to the browser? instead say issuer is not matched.